Show binding changes

causztic opened the pull request:
Currently on rails error pages, the change of the binding when clicking on the Trace links is not very visible to users.

This PR will add a notification div that will notify the user the current binding that the user is working with in the console.

It uses a simple callback with the anchor's innerHTML and displays it above the last prompt-box.

<img width="1440" alt="screen shot 2018-06-08 at 11 34 31 am" src="https://user-images.githubusercontent.com/2081441/41138058-59558984-6b11-11e8-89a7-e952ef54da86.png">

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -864,11 +864,23 @@ REPLConsole.prototype.scrollToBottom = function() {
-// Change the binding of the console
+/**
+ * Change the binding of the console
+ * Also updates a div to show that it has been changed.
```
This may change, maybe we won't win a lot by changing the comment here.

causztic replied:
Makes sense, I'll revert the change here

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -864,11 +864,23 @@ REPLConsole.prototype.scrollToBottom = function() {
+  var notification = document.createElement('div');
+  var _this = this;
+
+  notification.className = "console-notification";
```
```js
var notification = document.createElement('div');
notification.className = "console-notification";

var _this = this;
postRequest(url, params, function() {
```

Can we keep the flow like this here? The operations are "grouped" this way.

causztic replied:
sure!

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -864,11 +864,23 @@ REPLConsole.prototype.scrollToBottom = function() {
+
+  notification.className = "console-notification";
+  postRequest(url, params, function() {
+    notification.innerHTML = "Binding has changed to: " + callback();
```
I'm thinking whether we call this `Context has changed to:` because we cannot expect the web console users to know about bindings or the terminology around how code evaluation works in Ruby.

causztic replied:
I initially had it as **Context** as well but renamed it last minute. I guess I can change it back! I agree it sounds more "intuitive" for people using web-console.

gsamokovarov commented on lib/web_console/templates/_inner_console_markup.html.erb:
```diff
@@ -3,6 +3,8 @@
   </div>
-  <div class='console-inner'></div>
+  <div class='console-inner'>
+    <div class='console-notifications'></div>
```
Do we really need the extra layout here? What's so special about the notifications, they look like a regular console line to me.

causztic replied:
Initially I had the ```console-notifications``` set to be updated whenever the binding changes, but it will disappear out of the scrolling window when the commands exceeds the scrollable view. 

I have changed it to append before the lastChild of the ```prompt-box``` so that it remains in view, but I forgot to remove the unused ```console-notifications``` div. It is now removed!

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -864,11 +864,20 @@ REPLConsole.prototype.scrollToBottom = function() {
   var params = "frame_id=" + encodeURIComponent(frameId);
-  postRequest(url, params, callback);
+  var notification = document.createElement('div');
+  notification.className = "console-notification";
```
Do we need to just feed a new input line?

causztic replied:
Currently the default unclassed console text color is black so it won't be visible. I can remove the div but the console text color will need to be changed

causztic commented:
I have refactored the ```writeOutput``` and ```writeError``` into a ```write``` function accepting ```type``` and ```output```.

Also introduced a ```notification``` type for binding switches and used a complementary colour against the current error colour. 

Also made all colour codes consistent.

causztic commented:
Reverted the refactoring and added an extra prototype function for writeNotifications instead.

gsamokovarov commented on lib/web_console/templates/_inner_console_markup.html.erb:
```diff
@@ -3,6 +3,7 @@
     <div class='close-button button' title='close'>x</div>
   </div>
-  <div class='console-inner'></div>
+  <div class='console-inner'>
```
We may drop this diff change as well, e.g. leave it as before:

```
<div class='console-inner'></div>
```

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -595,13 +595,19 @@ REPLConsole.prototype.writeOutput = function(output) {
   this.newPromptBox();
   return consoleMessage;
-};
+}
```
This one too, leave the `};`, so we don't have this one pop-up in the diff.

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -595,13 +595,19 @@ REPLConsole.prototype.writeOutput = function(output) {
   addClass(consoleMessage, "error-message");
   return consoleMessage;
-};
+}
```
Same here.

gsamokovarov commented on lib/web_console/templates/console.js.erb:
```diff
@@ -595,13 +595,19 @@ REPLConsole.prototype.writeOutput = function(output) {
+  var consoleMessage = this.writeOutput(output);
+  addClass(consoleMessage, "notification-message");
+  return consoleMessage;
+}
```
Same here.
